cpp/src/parquet/column_writer.cc

```diff
@@ -1364,6 +1367,41 @@ class TypedColumnWriterImpl : public ColumnWriterImpl, public TypedColumnWriter<
                          int64_t num_levels, const ::arrow::Array& array,
                          ArrowWriteContext* context, bool maybe_parent_nulls);
 
+  template <typename ArrowType>
+  Status WriteArrowSerialize(const int16_t* def_levels, const int16_t* rep_levels,
+                             int64_t num_levels, const ::arrow::Array& array,
+                             ArrowWriteContext* ctx, bool maybe_parent_nulls);
+
+  Status WriteArrowZeroCopy(const int16_t* def_levels, const int16_t* rep_levels,
+                            int64_t num_levels, const ::arrow::Array& array,
+                            ArrowWriteContext* ctx, bool maybe_parent_nulls) {
+    const auto& data = static_cast<const ::arrow::PrimitiveArray&>(array);
+    const T* values = nullptr;
+    // The values buffer may be null if the array is empty (ARROW-2744)
+    if (data.values() != nullptr) {
+      values = reinterpret_cast<const T*>(data.values()->data()) + data.offset();
+    } else {
+      DCHECK_EQ(data.length(), 0);
+    }
```

Comment:
Can probably change this to:
```suggestion
    const T* values = data.data()->GetValues<T>(1);
```

Reply:
Updated.